### PR TITLE
IGNITE-16390

### PR DESCRIPTION
**Pull request by denis-chudov**

### CODE REVIEW

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/schema/SqlSchemaManagerImpl.java [hunk 1]
```diff
@@ -17,213 +17,376 @@
...
 
         newCalciteSchema.add("PUBLIC", new IgniteSchema("PUBLIC"));
 
-        igniteSchemas.forEach(newCalciteSchema::add);
-        externalCatalogs.forEach(newCalciteSchema::add);
-
-        calciteSchema = newCalciteSchema;
+        return CompletableFuture.allOf(
+            schemasVv.get(causalityToken).thenAccept(schemas -> schemas.forEach(newCalciteSchema::add)),
+            externalCatalogsVv.get(causalityToken).thenAccept(catalogs -> catalogs.forEach(newCalciteSchema::add))
```
Comment (vldpyatkov) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/schema/SqlSchemaManagerImpl.java:384]: I think, here previous token is required (previous from the causalityToken).

File: modules/core/src/main/java/org/apache/ignite/internal/causality/CausalityTokensLockManager.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+package org.apache.ignite.internal.causality;
+
+/**
+ * Lock manager that manages locks on causality tokens between multiple {@code VersionedValues}.
+ * When token is locked using some {@link VersionedValue} (see {@link VersionedValue#lock(long)}),
+ * it will be locked on every {@link VersionedValue} that shares the same {@link CausalityTokensLockManager}.
+ */
+// TODO https://issues.apache.org/jira/browse/IGNITE-16544
+public class CausalityTokensLockManager {
```
[the remark this answers is not in the document]
Comment (denis-chudov) [reply modules/core/src/main/java/org/apache/ignite/internal/causality/CausalityTokensLockManager.java]: done

File: modules/core/src/main/java/org/apache/ignite/internal/causality/VersionedValue.java [hunk 3]
```diff
@@ -0,0 +1,428 @@
...
+    /**
+     * Locks the given causality token to prevent its deletion from history of every {@link VersionedValue}
+     * that share the same {@link CausalityTokensLockManager} as this.
+     * This method works like read lock in read-write lock concept, allowing multiple {@link VersionedValue} to
+     * acquire lock on same token.
+     *
+     * @param causalityToken Causality token.
+     * @throws OutdatedTokenException If outdated token is passed as an argument.
+     */
+    public void lock(long causalityToken) throws OutdatedTokenException {
```
[the remark this answers is not in the document]
Comment (denis-chudov) [reply modules/core/src/main/java/org/apache/ignite/internal/causality/VersionedValue.java]: agree, removing along with lock manager

File: modules/core/src/main/java/org/apache/ignite/lang/PatchedMapView.java [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+import java.util.Set;
+import java.util.function.Function;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * This map view is constructed from a map and a single change applied to it, like put, remove or clear.
+ * It is not thread-safe.
+ */
+public class PatchedMapView<K, V> implements Map<K, V> {
```
[the remark this answers is not in the document]
Comment (denis-chudov) [reply modules/core/src/main/java/org/apache/ignite/lang/PatchedMapView.java:33]: It is needed to reduce the expenses of creating new maps on every configuration update. I will create another issue related to it.

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/schema/IgniteSchema.java [hunk 5]
```diff
@@ -29,15 +29,25 @@
 public class IgniteSchema extends AbstractSchema {
     private final String schemaName;
 
-    private final Map<String, InternalIgniteTable> tblMap = new ConcurrentHashMap<>();
+    private final Map<String, Table> tblMap;
```
[the remark this answers is not in the document]
Comment (denis-chudov) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/schema/IgniteSchema.java:32]: `InternalIgniteTable` extends `Table` and seems is not necessary here. In the same time, `#getTableMap` returns map of `<String, Table>`.

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/schema/SqlSchemaManagerImpl.java (hunk 1 above)
Comment (denis-chudov) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/schema/SqlSchemaManagerImpl.java:384]: rewrote it, with VersionedValue#update returning T it became more simple.

File: modules/runner/src/main/java/org/apache/ignite/internal/app/IgniteImpl.java [hunk 6]
```diff
@@ -197,21 +202,42 @@
...
+
         baselineMgr = new BaselineManager(
                 clusterCfgMgr,
                 metaStorageMgr,
                 clusterSvc
         );
 
+        Consumer<Consumer<Long>> registry = (c) -> {
+            clusterCfgMgr.configurationRegistry().listenUpdateStorageRevision(newStorageRevision -> {
+                LOG.info("Node: " + name);
```
[the remark this answers is not in the document]
Comment (denis-chudov) [reply modules/runner/src/main/java/org/apache/ignite/internal/app/IgniteImpl.java:225]: removed

File: modules/runner/src/main/java/org/apache/ignite/internal/app/IgniteImpl.java [hunk 7]
```diff
@@ -197,21 +202,42 @@
...
                 modules.distributed().polymorphicSchemaExtensions()
         );
 
+        Consumer<Consumer<Long>> storageRevisionUpdater = c -> clusterCfgMgr.configurationRegistry().listenUpdateStorageRevision(rev -> {
+            c.accept(rev);
+
+            return completedFuture(null);
+        });
+
+        Supplier<CompletableFuture<Long>> directMsRevisionSup = cfgStorage::lastRevision;
```
[the remark this answers is not in the document]
Comment (denis-chudov) [reply modules/runner/src/main/java/org/apache/ignite/internal/app/IgniteImpl.java]: done

**Pull request closed without merging**